File: Striver/Recursion/q14.java
package Striver.Recursion;

public class q14 {
    // print all combination sum of array elements which sum to target where each element can be used any number of times
    
    public static void printSubsequence(int[] arr, int k, int i, String ans) {
        if (i == arr.length) {
            if (k == 0) {
                System.out.println(ans);
            }
            return;
        }
        if (k - arr[i] >= 0) {
            printSubsequence(arr, k - arr[i], i, ans + arr[i] + " ");
        }
        printSubsequence(arr, k, i + 1, ans);
    }

    public static void main(String[] args) {
        int[] arr = { 2, 3, 5, 7 };
        int k = 10;
        printSubsequence(arr, k, 0, "");
    }
}

File: Striver/Recursion/q14.py
# print all combination sum of array elements which sum to target where each element can be used any number of times

def combinationSum(arr, target, index, path):
    if index == len(arr):
        if target == 0:
            print(path)
        return
    if target - arr[index] >= 0:
        combinationSum(arr, target - arr[index], index, path + [arr[index]])
    combinationSum(arr, target, index + 1, path)

arr = [2, 3, 5, 7]
combinationSum(arr, 7, 0, [])